File: app/src/androidTest/java/com/boost/espressotest/presentation/screen/main/view/MainActivityTest.java
package com.boost.espressotest.presentation.screen.main.view;


import android.support.test.espresso.intent.rule.IntentsTestRule;
import android.support.test.runner.AndroidJUnit4;
import android.widget.AutoCompleteTextView;

import com.boost.espressotest.R;
import com.boost.espressotest.presentation.screen.detail.view.DetailActivity;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.action.ViewActions.typeText;
import static android.support.test.espresso.assertion.ViewAssertions.doesNotExist;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.contrib.RecyclerViewActions.scrollToPosition;
import static android.support.test.espresso.intent.Intents.intending;
import static android.support.test.espresso.intent.matcher.IntentMatchers.hasExtra;
import static android.support.test.espresso.matcher.ViewMatchers.hasDescendant;
import static android.support.test.espresso.matcher.ViewMatchers.isAssignableFrom;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static com.boost.espressotest.matcher.RecyclerViewMatcher.atPosition;
import static org.hamcrest.core.IsNot.not;

@RunWith(AndroidJUnit4.class)
public class MainActivityTest {

    private static final int FIRST_ITEM = 0;
    private static final int ITEM_TO_SCROLL = 27;

    @Rule
    public IntentsTestRule<MainActivity> mMainActivityIntentRule = new IntentsTestRule<>(MainActivity.class);

    @Test
    public void scrollToItem_verifyItemDisplayed() {
        onView(withId(R.id.rv_products)).perform(scrollToPosition(ITEM_TO_SCROLL));
        String name = mMainActivityIntentRule.getActivity().getResources().getString(R.string.mock_name) + String.valueOf(ITEM_TO_SCROLL);
        onView(withText(name)).check(matches(isDisplayed()));

        onView(withId(R.id.rv_products)).perform(scrollToPosition(ITEM_TO_SCROLL / 2));
        String name2 = mMainActivityIntentRule.getActivity().getResources().getString(R.string.mock_name) + String.valueOf(ITEM_TO_SCROLL / 2);
        onView(withText(name2)).check(matches(isDisplayed()));
    }

    @Test
    public void searchItemByName_verifyItemIsOnTopAndSingle() {
        onView(withId(R.id.sv_search)).perform(click());
        onView(isAssignableFrom(AutoCompleteTextView.class)).perform(typeText(String.valueOf(ITEM_TO_SCROLL)));

        String name = mMainActivityIntentRule.getActivity().getResources().getString(R.string.mock_name) + String.valueOf(ITEM_TO_SCROLL);

        onView(withText(String.valueOf(ITEM_TO_SCROLL - 1))).check(doesNotExist());
        onView(withId(R.id.rv_products)).check(matches(atPosition(FIRST_ITEM, hasDescendant(withText(name)))));
    }

    @Test
    public void searchExampleText_verifyAnyItemIsDisplayed() {
        onView(withId(R.id.sv_search)).perform(click());
        onView(isAssignableFrom(AutoCompleteTextView.class)).perform(typeText("test empty search"));
        onView(withId(R.id.rv_products)).check(matches(not(atPosition(FIRST_ITEM, hasDescendant(withId(R.id.tv_product_title))))));
    }

    @Test
    public void navigateToDetailedScreen_verifyNavigation() {
        String name = mMainActivityIntentRule.getActivity().getResources().getString(R.string.mock_name) + String.valueOf(FIRST_ITEM);

        onView(withText(name)).perform(click());

        intending(hasExtra(DetailActivity.ARG_PRODUCT_ID, FIRST_ITEM));
    }
}

File: app/src/androidTest/java/com/boost/espressotest/presentation/screen/main/view/MainActivityTest.kt
package com.boost.espressotest.presentation.screen.main.view

import android.view.View
import android.widget.AutoCompleteTextView
import androidx.recyclerview.widget.RecyclerView
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.typeText
import androidx.test.espresso.assertion.ViewAssertions.doesNotExist
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.contrib.RecyclerViewActions.scrollToPosition
import androidx.test.espresso.intent.Intents.intending
import androidx.test.espresso.intent.matcher.IntentMatchers.hasExtra
import androidx.test.espresso.intent.rule.IntentsTestRule
import androidx.test.espresso.matcher.ViewMatchers.*
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.boost.espressotest.R
import com.boost.espressotest.matcher.RecyclerViewMatcher.atPosition
import com.boost.espressotest.presentation.screen.detail.view.DetailActivity
import org.hamcrest.core.IsNot.not
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class MainActivityTest {

    companion object {
        private const val FIRST_ITEM = 0
        private const val ITEM_TO_SCROLL = 27
    }

    @Rule @JvmField
    var mMainActivityIntentRule = IntentsTestRule(MainActivity::class.java)

    @Test
    fun openScreen_verifyTitleIsCorrect() {
        val title = mMainActivityIntentRule.activity.resources.getString(R.string.main_title)
        onView(withId(R.id.tv_title)).check(matches(withText(title)))
    }

    @Test
    fun scrollToItem_verifyItemDisplayed() {
        onView(withId(R.id.rv_products)).perform(scrollToPosition<RecyclerView.ViewHolder>(ITEM_TO_SCROLL))
        val name = mMainActivityIntentRule.activity.resources.getString(R.string.mock_name) + ITEM_TO_SCROLL.toString()
        onView(withText(name)).check(matches(isDisplayed()))

        onView(withId(R.id.rv_products)).perform(scrollToPosition<RecyclerView.ViewHolder>(ITEM_TO_SCROLL / 2))
        val name2 = mMainActivityIntentRule.activity.resources.getString(R.string.mock_name) + (ITEM_TO_SCROLL / 2).toString()
        onView(withText(name2)).check(matches(isDisplayed()))
    }

    @Test
    fun searchItemByName_verifyItemIsOnTopAndSingle() {
        onView(withId(R.id.sv_search)).perform(click())
        onView(isAssignableFrom(AutoCompleteTextView::class.java)).perform(typeText(ITEM_TO_SCROLL.toString()))

        val name = mMainActivityIntentRule.activity.resources.getString(R.string.mock_name) + ITEM_TO_SCROLL.toString()

        onView(withText((ITEM_TO_SCROLL - 1).toString())).check(doesNotExist())
        onView(withId(R.id.rv_products)).check(matches(atPosition(FIRST_ITEM, hasDescendant(withText(name)))))
    }

    @Test
    fun searchExampleText_verifyAnyItemIsDisplayed() {
        onView(withId(R.id.sv_search)).perform(click())
        onView(isAssignableFrom(AutoCompleteTextView::class.java)).perform(typeText("test empty search"))
        onView(withId(R.id.rv_products)).check(matches(not<View>(atPosition(FIRST_ITEM, hasDescendant(withId(R.id.tv_product_title))))))
    }

    @Test
    fun navigateToDetailedScreen_verifyNavigation() {
        val name = mMainActivityIntentRule.activity.resources.getString(R.string.mock_name) + FIRST_ITEM.toString()

        onView(withText(name)).perform(click())

        intending(hasExtra(DetailActivity.ARG_PRODUCT_ID, FIRST_ITEM))
    }
}